tonic-reflection/src/server.rs

```diff
@@ -0,0 +1,323 @@
+use crate::proto::server_reflection_request::MessageRequest;
+use crate::proto::server_reflection_response::MessageResponse;
+use crate::proto::server_reflection_server::{ServerReflection, ServerReflectionServer};
+use crate::proto::{
+    FileDescriptorResponse, ListServiceResponse, ServerReflectionRequest, ServerReflectionResponse,
+    ServiceResponse,
+};
+use prost::Message;
+use prost_types::{
+    DescriptorProto, EnumDescriptorProto, FieldDescriptorProto, FileDescriptorProto,
+};
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::stream::StreamExt;
+use tokio::sync::mpsc;
+use tonic::{Request, Response, Status, Streaming};
+
+/// Represents an error in the construction of a gRPC Reflection Service.
+#[derive(Debug, Error)]
+pub enum Error {
+    /// An error was encountered decoding a `prost_types::FileDescriptorProto` from a buffer.
+    #[error("error decoding FileDescriptorProto from buffer")]
+    DecodeError(#[from] prost::DecodeError),
+    /// A `prost_types::FileDescriptorProto`
+    #[error("invalid FileDescriptorProto - {0}")]
+    InvalidFileDescriptorProto(String),
+}
+
+/// A builder used to construct a gRPC Reflection Service.
+#[derive(Debug)]
+pub struct Builder {
+    file_descriptors: Vec<FileDescriptorProto>,
+    encoded_file_descriptors: Vec<&'static [u8]>,
+    include_reflection_service: bool,
+
+    service_names: Vec<String>,
+    symbols: HashMap<String, Arc<FileDescriptorProto>>,
+}
+
+impl Builder {
+    /// Create a new builder that can configure a gRPC Reflection Service.
+    pub fn configure() -> Self {
+        Builder {
+            file_descriptors: Vec::new(),
+            encoded_file_descriptors: Vec::new(),
+            include_reflection_service: true,
+
+            service_names: Vec::new(),
+            symbols: HashMap::new(),
+        }
+    }
+
+    /// Registers an instance of `prost_types::FileDescriptorProto` with the gRPC Reflection
+    /// Service builder.
+    pub fn register_file_descriptor(mut self, file_descriptor: FileDescriptorProto) -> Self {
+        self.file_descriptors.push(file_descriptor);
+        self
+    }
+
+    /// Registers a byte slice containing an encoded `prost_types::FileDescriptorProto` with
+    /// the gRPC Reflection Service builder.
+    pub fn register_encoded_file_descriptor(
+        mut self,
+        encoded_file_descriptor: &'static [u8],
```

Comment:
why does this need to be static? Would it better to accept a `&[u8]` and box it?